### ISSUE TITLE
RTCP: All RTCP packets need to be compound packets

### ISSUE THREAD

**Issue by dadosch** [ultravideo/uvgRTP#91]
As RFC 3550 describes:
>    Thus, all RTCP packets MUST be sent in a compound packet of at least
   two individual packets, with the following format:

> …

>   SR or RR:  The first RTCP packet in the compound packet MUST
      always be a report packet to facilitate header validation as
      described in Appendix A.2.  This is true even if no data has been
      sent or received, in which case an empty RR MUST be sent, and even
      if the only other RTCP packet in the compound packet is a BYE.

> …

>   SDES:  An SDES packet containing a CNAME item MUST be included
      in each compound RTCP packet, except as noted in Section 9.1.
      Other source description items MAY optionally be included if
      required by a particular application, subject to bandwidth
      constraints (see Section 6.3.9).

Receive Reports and Send Reports must be in a RTCP Compund Packet together with a SDES packet (CNAME). Consequently, Wireshark marks RTCP Reports without a SDES CNAME as invalid.

(Related to https://github.com/ultravideo/uvgRTP/issues/51)

**Comment by jrsnen** [comment]
Thank you for bringing this to our attention. This is confirmed in: https://datatracker.ietf.org/doc/html/rfc3550#section-6.1

### Relevant parts in code

Receiving compound packets:
https://github.com/ultravideo/uvgRTP/blob/9ef02a72f891702f6cfb281613426bccc67c109f/src/rtcp.cc#L756

Generating reports (function name should probably be changed):
https://github.com/ultravideo/uvgRTP/blob/9ef02a72f891702f6cfb281613426bccc67c109f/src/rtcp.cc#L1172

SDES is sent using this function:
https://github.com/ultravideo/uvgRTP/blob/9ef02a72f891702f6cfb281613426bccc67c109f/src/rtcp.cc#L1271

There also seem to be issues in that function.

Unsure if the API needs changes, so packets can be grouped together, or if it is enough that they are received at the same time.

**Comment by jrsnen** [comment]
I read the RFC and noticed that all RTCP packets should be in compound packets with SR and RR. I edited the title.

**Comment by jrsnen** [comment]
I made some progress on compound packets, but #137 should be fixed before they can be brought into use.

**Comment by jrsnen** [comment]
SDES compound packet implemented in d06911e07d3155b715b5ef76a4597399fef8bedf

APP and BYE compound packets implemented in f4461f6dc3642da54b762c40e691206e845c770c

uvgRTP now uses exclusively compound packets, as RFC 3550 specifies. 

SDES seems to work without issues.

There are still issues with BYE, but they should be resolved in #111

I didn't test APP, but it should work by sending one packet per APP name together with each SR/RR. This info may interest @wowaser 

This issue is fixed.

**Comment by wowaser** [comment]
>  but it should work by sending one packet per APP name together with each SR/RR 

Does this mean that when I call **rtcp->send_app_packet**  the packet is queued, and then glued together with an SR/RR? So the max frequency of APP packets is *1/RTCP interval*?  What does "one packet per APP name" mean?

**Comment by jrsnen** [comment]
@wowaser Yes and yes, only one APP packet per SR/RR as the RTCP protocol is designed to be a low bandwidth protocol compared to RTP. I made it so that all app packets are queued per name, waiting for the next SR/RR to be sent and taken off the queue. It will give you a warning if you send too fast. 

If you have ideas on how to improve the API on this, feel free to make a PR. I can maybe see situations where getting the info when the APP packet is sent through a callback might be beneficial.

APP packet has the name field, so I made it so that each different name gets one packet in SR/RR compound packet (as their purpose should be different). It is possible that each subtype should also get its own packet, but I was too lazy to figure it out.

All this was made so uvgRTP is compliant with RFC 3550.
